### PR TITLE
Add RCNNLossLayer, RCNNDetectionLayer for Faster(er) R-CNN

### PR DESCRIPTION
**Pull request by guoshengCS**
Add RCNNLossLayer, RCNNDetectionLayer for Faster(er) R-CNN.

### CODE REVIEW

File: paddle/gserver/layers/RCNNDetectionLayer.cpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+  // <batchIdx, <classIdx, <(score, box)>>>
+  std::map<size_t,
+           std::map<size_t, std::vector<std::pair<real, NormalizedBBox>>>>
+      allDecodedBBoxes;
+  for (size_t n = 0; n < roiNum; ++n) {
+    int batchIdx = *(roisData + n * roiDim);
+    std::vector<real> roiLocData(4);  // RoI location
+    for (size_t j = 0; j < 4; ++j) {
+      roiLocData[j] = *(roisData + n * roiDim + 1 + j);
+    }
```
Comment (qingqing01) [inline paddle/gserver/layers/RCNNDetectionLayer.cpp]:
```
int batchIdx = *(roisData + n * roiDim);
std::vector<real> roiLocData(4);  // RoI location
for (size_t j = 0; j < 4; ++j) {
  roiLocData[j] = *(roisData + n * roiDim + 1 + j);
}
```
==>
```
roisData += roiDim;
int batchIdx = *roisData;
std::vector<real> roiLocData(roisData+ 1, roisData+ 5);
```

std::vector的初始化：http://www.cplusplus.com/reference/vector/vector/vector/

File: paddle/gserver/layers/RCNNDetectionLayer.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    // location predictions for each class
+    for (size_t c = 0; c < numClasses_; ++c) {
+      if (c == backgroundId_) continue;
+      std::vector<real> predLocData(4);
+      for (size_t j = 0; j < 4; ++j) {
+        predLocData[j] = *(locPredData + n * numClasses_ * 4 + c * 4 + j);
+      }
+      real predConfData = *(confPredData + n * numClasses_ + c);
```
Comment (qingqing01) [inline paddle/gserver/layers/RCNNDetectionLayer.cpp]: 同样代码可以短一些：
```c++
locPredData += numClasses_ * 4;
for (size_t c = 0; c < numClasses_; ++c) {
  if (c == backgroundId_) continue;
  std::vector<real> predLocData(locPredData + c * 4, locPredData + c * 4 + 4);
  real predConfData = *(confPredData + c);
  // ...
```

File: paddle/gserver/layers/RCNNDetectionLayer.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * The detection output layer to generate bounding boxes in Fast R-CNN.
+ * This layer applies Non-maximum suppression to all predicted bounding
+ * boxes and keeps the Top-K bounding-boxes.
+ * - Input: This layer needs three input layers: The first input layer
+ *          contains the prior-box data. The rest two input layers are
+ *          layers for generating bounding-box location offset and the
+ *          classification confidence.
+ * - Output: The predict bounding boxes.
```
Comment (qingqing01) [inline paddle/gserver/layers/RCNNDetectionLayer.h]: The predicted bounding boxes.

File: python/paddle/trainer/config_parser.py [hunk 4]
```diff
@@ -1781,6 +1781,39 @@ def __init__(self, name, inputs, size, input_num, num_classes,
         self.config.size = size
 
 
+@config_layer('rcnn_loss')
+class RCNNLossLayer(LayerBase):
+    def __init__(self, name, inputs, loss_ratio, num_classes, background_id=0):
```
Comment (qingqing01) [inline python/paddle/trainer/config_parser.py]: ```, **xargs):```

File: python/paddle/trainer_config_helpers/layers.py [hunk 5]
```diff
@@ -1249,6 +1253,101 @@ def detection_output_layer(input_loc,
...
 
 
+@wrap_name_default("rcnn_loss")
+def rcnn_loss_layer(rois,
+                    input_loc,
+                    input_conf,
+                    loss_ratio,
+                    num_classes,
+                    background_id=0,
+                    name=None):
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: it's better to add some comments for the difference between `rnn_loss` and `multibox_loss`, if you understand.

**Review by qingqing01 — COMMENTED** [review]
@pkuyym Please help to review.

**Comment by luotao1** [comment]
感谢您给PaddlePaddle贡献代码。由于Paddle V1/V2版本已不再维护，相关代码也已从develop分支上删除，因此关闭您的PR，欢迎您向Paddle最新版-Fluid贡献代码。
Thanks for contributing to PaddlePaddle! Since V1/V2 will not be maintained anymore, and related codes have been deleted from develop branch as well, we close this PR. Welcome to contribute to Fluid——the latest version of PaddlePaddle.

**Pull request closed without merging**